### ISSUE TITLE
Image preprocessing parameters

### ISSUE THREAD

**Issue by lkforward** [qubvel/segmentation_models.pytorch#220]
The function `preprocess_input()`, in encoders/_preprocessing.py, takes 'mean' and 'std' as parameters and apply the normalization on the data in the following way:

```
    if mean is not None:
        mean = np.array(mean)
        x = x - mean

    if std is not None:
        std = np.array(std)
        x = x / std
```

In my opinion, the mean/std here should be the statistics for the training dataset (data-specific). However, according to `get_preprocessing_params()`, in encoders/\_\_init\_\_.py, the mean/std are determined by the pretrained model, which depends on the training data used in the pretrained model. 

Just wonder, is there any reason why we do it based on the pretrained model?

**Comment by howard-mahe** [comment]
Yes. The pretrained models was trained for a given mean/std in the preprocessing stage. If you input an ImageNet image to your encoder, it has to achieve the expected predictions.

However, in the specific case where you fine-tune your encoder (i.e. as part of a segmentation model) on a dataset not being ImageNet, indeed your data statistics is likely not the same as the ImageNet data statistics. In that case, it does not matter much whether you use ImageNet mean/std, your dataset mean/std or even mean=0.5 and std=0.5 (which maps [0,1] to [-1,1]) because in any cases, the weights of the first layers of your encoder will have to adapt to the new data statistics.

**Comment by lkforward** [comment]
@howard-mahe Thank you very much for the clarification, and that is very helpful. 

My only concern is, if we use the hard-coded mean / std from the pre-trained model, it might bring the distribution of the new training dataset further away from the data the pretrained model was fitted into, and might not be good for optimization. 

Taking a step back, the reason we want to do this preprocessing is to normalize the data and make it easier for the optimization algorithm (i.e, gradient descent) to find the optimum. And in principle, if we want to normalize a dataset, we apply the transformation: x ' = (x - mean) / std, where mean and std are from the same dataset. 

Anyway, I agree that whatever we choose, the weights of the first layers of the encoder will have to adapt to the new data statistics. We just want to make a careful choice so the distribution of the dataset works better for the optimizer (and its associated hyperparameters) and the pretrained weights (as the initial values for the fine-tuning).
